casper/src/test/scala/coop/rchain/casper/batch2/BlockValidationLogicSpec.scala

```diff
@@ -0,0 +1,150 @@
+import coop.rchain.crypto.{PrivateKey, PublicKey}
+import coop.rchain.models.BlockVersion
+import coop.rchain.models.blockImplicits.{arbBlockMessage, getRandomBlock}
+import coop.rchain.models.syntax._
+import org.scalacheck.Arbitrary
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+import org.scalatestplus.scalacheck.ScalaCheckPropertyChecks
+
+class BlockValidationLogicSpec extends AnyFlatSpec with Matchers with ScalaCheckPropertyChecks {
+
+  implicit private val aBlock: Arbitrary[BlockMessage] = arbBlockMessage
+
+  "Block version validation" should "allow supported versions" in {
+    forAll { (block: BlockMessage, version: Int) =>
+      val blockWithVersion = block.copy(version = version)
+
+      // Expected one of hard-coded block versions supported by this version of RNode software
+      val expectedValid = BlockVersion.Supported.contains(version)
+      // Actual validation
+      val actualValid = BlockValidationLogic.version(blockWithVersion)
+
+      actualValid shouldBe expectedValid
+    }
+  }
+
+  "Block hash format validation" should "fail on invalid hash" in {
+    forAll { (block: BlockMessage) =>
+      val hash           = ProtoUtil.hashBlock(block)
+      val blockValidHash = block.copy(blockHash = hash)
+
+      // Test valid block hash
+      val hashValid = BlockValidationLogic.blockHash(blockValidHash)
+
+      hashValid shouldBe true
+
+      val blockInValidHash = block.copy(blockHash = ByteString.copyFromUtf8("123"))
+
+      // Test invalid block hash
+      val hashInValid = BlockValidationLogic.blockHash(blockInValidHash)
+
+      hashInValid shouldBe false
+    }
+  }
+
+  "Field format validation" should "succeed on a valid block and fail on empty fields" in {
+    val (privateKey, publicKey) = Secp256k1.newKeyPair
+    val genesis                 = signedBlock(privateKey, publicKey)
+
+    BlockValidationLogic.formatOfFields(genesis) shouldBe true
+    BlockValidationLogic.formatOfFields(genesis.copy(blockHash = ByteString.EMPTY)) shouldBe false
+    BlockValidationLogic.formatOfFields(genesis.copy(sig = ByteString.EMPTY)) shouldBe false
+    BlockValidationLogic.formatOfFields(genesis.copy(sigAlgorithm = "")) shouldBe false
+    BlockValidationLogic.formatOfFields(genesis.copy(shardId = "")) shouldBe false
+    BlockValidationLogic.formatOfFields(genesis.copy(postStateHash = ByteString.EMPTY)) shouldBe false
+  }
+
+  // TODO: This test doesn't make sense because the signing is done by the algorithm in the ValidatorIdentity.
+  //  The algorithm specified for the block is not used
+  "Block signature validation" should "return false on unknown algorithms" in {
+    val blockUnknownAlg = getRandomBlock().copy(sigAlgorithm = "unknownAlgorithm")
+    val blockRsaAlg     = getRandomBlock().copy(sigAlgorithm = "RSA")
+
+    BlockValidationLogic.blockSignature(blockUnknownAlg) shouldBe false
+    BlockValidationLogic.blockSignature(blockRsaAlg) shouldBe false
+  }
```

Comment:
It seems previously there was misunderstanding what part of the block data should be hashed and what is signature data. We now need to fix this is steps.
https://github.com/rchain/rchain/blob/efa33eb02b291268eb937c03fd7ec1dca90e574d/casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala#L62-L67

Reply:
I'm newby to cryptography. It's strange to me that we are signing hash of the block instead of the entire block. And why duplicate `sigAlgorithm` if it is in `ValidatorIdentity`. What can you suggest for this test?

Reply:
Signature algorithm usually expect some small data as an input so hashing the data before signing is a common practice.

---

ValidatorIdentity contains current signature algorithm for the node to create new blocks. In some moment signature algorithm can be changed so we need to be able to validate blocks created with the previous version.

---

For test you can also use generative style and prove that other part of the data of the block are not relevant for validation of the signature.
Later we can improve fields validation and be more specific in e.g. length of the signature or hashes and also validate signature algorithm separately from validating the signature.